internal/core/items/text.rs

```diff
@@ -801,30 +801,31 @@ impl Item for TextInput {
                     }
                 }
 
+                if self.read_only() || event.modifiers.control {
+                    return KeyEventResult::EventIgnored;
+                }
+
+                // save real anchor/cursor for undo/redo
+                let (real_cursor, real_anchor) = {
+                    let text = self.text();
+                    (self.cursor_position(&text), self.anchor_position(&text))
+                };
+
                 let input_type = self.input_type();
                 if input_type == InputType::Number
                     && !event.text.as_str().chars().all(|ch| ch.is_ascii_digit())
                 {
                     return KeyEventResult::EventIgnored;
-                }
-                if input_type == InputType::Decimal {
-                    let text = self.text().clone() + event.text.as_str();
+                } else if input_type == InputType::Decimal {
+                    let text = self.text();
+                    let text =
+                        [&text[..real_anchor], event.text.as_str(), &text[real_cursor..]].concat();
```

Comment:
This code assumes that the anchor byte offset is <= the cursor byte offset. I think that if you select text to the left, by pressing shift and cursor left, that assumption will be incorrect. Perhaps it would make sense here to call `selection_anchor_and_cursor` and that way always get the correctly ordered offsets and thus safely replace the selected text with the entered text and then do the validation.